libfwupdplugin/fu-cabinet.c

```diff
@@ -129,6 +129,88 @@ fu_cabinet_get_file_by_name (FuCabinet *self, const gchar *basename)
 	return NULL;
 }
 
+/**
+ * fu_cabinet_add_file:
+ * @self: A #FuCabinet
+ * @basename: filename
+ * @data: #GBytes
+ *
+ * Adds a file to the silo.
+ *
+ * Since: 1.6.0
+ **/
+void
+fu_cabinet_add_file (FuCabinet *self, const gchar *basename, GBytes *data)
+{
+	GPtrArray *folders;
+	GCabFile *gcab_file_old;
+	g_autoptr(GCabFolder) gcab_folder = NULL;
+	g_autoptr(GCabFile) gcab_file = NULL;
+
+	g_return_if_fail (FU_IS_CABINET (self));
+	g_return_if_fail (basename != NULL);
+	g_return_if_fail (data != NULL);
+
+	/* existing file? */
+	gcab_file_old = fu_cabinet_get_file_by_name (self, basename);
+	if (gcab_file_old != NULL) {
```

Comment:
I wonder if you want an argument in this function to control overwrite behavior. I don't think it's always expected.

Reply:
I'd actually argue the opposite; `gcab_folder_add_file()` allows you to construct a perfectly valid .cab file with two GCabFile with the same name! This means you open it in the archive manager (e.g. file-roller) and you see two things with the same filename, but depending on what you use to decompress it you might get zero, 1 or 2 files! I think given we have `fu_cabinet_get_file()` there's not a scenario where you wouldn't want to overwrite an existing file with the same name.